src/ring_theory/witt_vector/basic.lean

```diff
@@ -0,0 +1,238 @@
+
+* `witt_vector.map`: lifts a ring homomorphism `R →+* S` to a ring homomorphism `𝕎 R →+* 𝕎 S`.
+* `witt_vector.ghost_component n x`: evaluates the `n`th Witt polynomial
+  on the first `n` coefficients of `x`, producing a value in `R`.
+  This is a ring homomorphism.
+* `witt_vector.ghost_map`: a ring homomorphism `𝕎 R →+* (ℕ → R)`, obtained by packaging
+  all the ghost components together.
+  If `p` is invertible in `R`, then the ghost map is an equivalence,
+  which we use to define the ring operations on `𝕎 R`.
+* `witt_vector.comm_ring`: the ring structure induced by the ghost components.
+
+## Notation
+
+We use notation `𝕎 R`, entered `\bbW`, for the Witt vectors over `R`.
+
+## Implementation details
+
+As we prove that the ghost components respect the ring operations, we face a number of repetitive
+proofs. To avoid duplicating code we factor these proofs into a custom tactic, only slightly more
+powerful than a tactic macro. This tactic is not particularly useful outside of its applications
+in this file.
+-/
+
+noncomputable theory
+
+open mv_polynomial function
+
+open_locale big_operators
+
+local attribute [semireducible] witt_vector
+
+variables {p : ℕ} {R S T : Type*} [hp : fact p.prime] [comm_ring R] [comm_ring S] [comm_ring T]
+variables {α : Type*} {β : Type*}
+
+local notation `𝕎` := witt_vector p -- type as `\bbW`
+open_locale witt
+
+namespace witt_vector
+
+/-- `f : α → β` induces a map from `𝕎 α` to `𝕎 β` by applying `f` componentwise.
+If `f` is a ring homomorphism, then so is `f`, see `witt_vector.map f`. -/
+def map_fun (f : α → β) : 𝕎 α → 𝕎 β := λ x, f ∘ x
+
+namespace map_fun
+
+lemma injective (f : α → β) (hf : injective f) : injective (map_fun f : 𝕎 α → 𝕎 β) :=
+λ x y h, funext $ λ n, hf $ by exact congr_fun h n
+
+lemma surjective (f : α → β) (hf : surjective f) : surjective (map_fun f : 𝕎 α → 𝕎 β) :=
+λ x, ⟨λ n, classical.some $ hf $ x n,
+by { funext n, dsimp [map_fun], rw classical.some_spec (hf (x n)) }⟩
+
+variables (f : R →+* S) (x y : 𝕎 R)
+
+/-- Auxiliary tactic for showing that `map_fun` respects the ring operations. -/
+meta def witt_map : tactic unit :=
+`[funext n,
+  show f (aeval _ _) = aeval _ _,
+  rw map_aeval,
+  apply eval₂_hom_congr (ring_hom.ext_int _ _) _ rfl,
+  ext ⟨i, k⟩,
+  fin_cases i; refl]
+
+include hp
+
+/- We do not tag these lemmas as `@[simp]` because the will be bundled in `map` later on. -/
+
+lemma zero : map_fun f (0 : 𝕎 R) = 0 := by witt_map
+
+lemma one : map_fun f (1 : 𝕎 R) = 1 := by witt_map
+
+lemma add : map_fun f (x + y) = map_fun f x + map_fun f y := by witt_map
+
+lemma mul : map_fun f (x * y) = map_fun f x * map_fun f y := by witt_map
+
+lemma neg : map_fun f (-x) = -map_fun f x := by witt_map
+
+end map_fun
+
+end witt_vector
+
+section tactic
+setup_tactic_parser
+open tactic
+
+/-- An auxiliary tactic for proving that `ghost_fun` respects the ring operations. -/
+meta def tactic.interactive.ghost_fun (φ fn : parse parser.pexpr) : tactic unit :=
```

Comment:
Can this be renamed to `ghost_fun_tac` or something? Just reading this on github, I can't work out which if any of the references to `ghost_fun` below refer to this.

Reply:
Good idea. Did the same thing for the `witt_map` tactic.